From 83a25308fc31136014d2b0eb67d07cc1e81fa916 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Tue, 16 May 2023 11:21:40 +0800
Subject: [PATCH 030/263] [Huawei] mm/mem_reliable: Fallback to dpool if
 reliable memory is not enough

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188703

--------------------------------

Fallback to dpool if reliable memory is not enough.

Kswapd and driect reclaim will not be trigger. Since the later
normal memory allocation can trigger this, there is no problem
here.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/dynamic_hugetlb.h |  5 +++++
 mm/dynamic_hugetlb.c            |  2 +-
 mm/page_alloc.c                 | 28 ++++++++++++++++++++++++++++
 3 files changed, 34 insertions(+), 1 deletion(-)

diff --git a/include/linux/dynamic_hugetlb.h b/include/linux/dynamic_hugetlb.h
index 31f81afa6d88..9f42065890b2 100644
--- a/include/linux/dynamic_hugetlb.h
+++ b/include/linux/dynamic_hugetlb.h
@@ -133,6 +133,7 @@ struct page *alloc_huge_page_from_dhugetlb_pool(struct hstate *h, struct dhugetl
 						bool need_unreserved);
 void free_huge_page_to_dhugetlb_pool(struct page *page, bool restore_reserve);
 bool page_belong_to_dynamic_hugetlb(struct page *page);
+bool should_allocate_from_dhugetlb_pool(gfp_t gfp_mask);
 
 int dpool_init(struct dpool_info *arg);
 void dpool_report_meminfo(struct seq_file *m);
@@ -209,6 +210,10 @@ bool page_belong_to_dynamic_hugetlb(struct page *page)
 {
 	return false;
 }
+static inline bool should_allocate_from_dhugetlb_pool(gfp_t gfp_mask)
+{
+	return false;
+}
 #endif
 
 struct dpool_info {};
diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index e72d9a0812f5..72ff716f96c5 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -507,7 +507,7 @@ int task_has_mem_in_hpool(struct task_struct *tsk)
 	return hpool ? -EPERM : 0;
 }
 
-static bool should_allocate_from_dhugetlb_pool(gfp_t gfp_mask)
+bool should_allocate_from_dhugetlb_pool(gfp_t gfp_mask)
 {
 	gfp_t gfp = gfp_mask & GFP_HIGHUSER_MOVABLE;
 
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index f159b1e707b8..2145a67894b7 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -4742,6 +4742,18 @@ static inline void mem_reliable_fallback_slowpath(gfp_t gfp_mask,
 		return;
 	}
 }
+
+static inline bool mem_reliable_fallback_dhugetlb(gfp_t gfp_mask)
+{
+	if (!reliable_allow_fb_enabled())
+		return false;
+
+	if ((gfp_mask & GFP_RELIABLE) &&
+	    should_allocate_from_dhugetlb_pool(gfp_mask & ~GFP_RELIABLE))
+		return true;
+
+	return false;
+}
 #else
 static inline struct zone *mem_reliable_fallback_zone(gfp_t gfp_mask,
 						      struct alloc_context *ac)
@@ -4750,6 +4762,10 @@ static inline struct zone *mem_reliable_fallback_zone(gfp_t gfp_mask,
 }
 static inline void mem_reliable_fallback_slowpath(gfp_t gfp_mask,
 						  struct alloc_context *ac) {}
+static inline bool mem_reliable_fallback_dhugetlb(gfp_t gfp_mask)
+{
+	return false;
+}
 #endif
 
 static inline struct page *
@@ -5410,6 +5426,18 @@ struct page *__alloc_pages(gfp_t gfp, unsigned int order, int preferred_nid,
 	if (likely(page))
 		goto out;
 
+	/*
+	 * Fallback to dpool if mirrored momory is not enough.
+	 *
+	 * Kswapd and driect reclaim will not be trigger, since the later
+	 * normal memory allocation can trigger this, there is no problem
+	 * here.
+	 */
+	if (mem_reliable_fallback_dhugetlb(gfp)) {
+		gfp &= ~GFP_RELIABLE;
+		goto retry;
+	}
+
 	/*
 	 * Apply scoped allocation constraints. This is mainly about GFP_NOFS
 	 * resp. GFP_NOIO which has to be inherited for all allocation requests
-- 
2.17.1

